Makefile

```diff
@@ -64,23 +61,122 @@ package: test
 	fi; \
 	}
 
+
 .PHONY: test-all
 test-all: build package
 
 .PHONY: ci-test
 ci-test:
-	# run tests for all ghc versions
-	# TODO: figure out a way to run even if in nix-shell
+	# run tests for all ghc versions given in different stack/*.yaml
 	{ \
 	set -e; \
-	if [ "$$IN_NIX_SHELL" ]; then \
-		echo "Invoked from within nix-shell; exiting"; \
-		exit 1; \
-	else \
-		for e in $(GHCS); do \
-			echo "Testing GHC $$e"; \
-			nix-shell shell-$$e.nix --command "make test-all"; \
-			echo "DONE GHC $$e"; \
-		done; \
-	fi; \
+	for stack_yaml in $$(ls stack); do \
+		echo "Testing $$stack_yaml"; \
+		stack --stack-yaml="stack/$$stack_yaml" build; \
+		stack --stack-yaml="stack/$$stack_yaml" test; \
+	done; \
 	}
+
+.PHONY: stack-build
+stack-build:
+	stack build
+
+.PHONY: stack-test
+stack-test: stack-build
+	stack test
+
+.PHONY: stack-bench
+stack-bench: stack-build stack-test # only benchmark tested program
+	stack bench
+
+
+
+.PHONY: stack-8.4.1-build
+stack-8.4.1-build:
```

Comment:
@power-fungus would it help reduce code duplication if the build/test targets used an environment variable (corresponding to ghc version) to run the right `stack` command? then we won't need these specialized commands and just need to set the right ghc version preference when invoking `make`. if no ghc version preference is provided, we may also default to latest one.